cpp/include/cuspatial/trajectory.hpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/types.hpp>
+#include <memory>
+#include <rmm/mr/device/default_memory_resource.hpp>
+
+namespace cuspatial {
+namespace experimental {
+
+/**
+ * @brief Derive trajectories from sorted object ids.
+ *
+ * Groups the input object ids to determine unique trajectories. Returns a
+ * table with the trajectory ids, the number of objects in each trajectory,
+ * and the offset position of the first object for each trajectory in the
+ * input object ids column.
+ *
+ * @param[in] id column of object (e.g., vehicle) ids
+ * @param[in] mr The optional resource to use for all allocations
+ *
+ * @return a sorted table with the following three int32 columns:
+ *   * trajectory id - the unique ids from the input object ids column
+ *   * trajectory length - the number of objects in the derived trajectories
+ *   * trajectory offset - the cumulative sum of start positions for each group
+ */
+std::unique_ptr<cudf::experimental::table> derive_trajectories(
```

Comment:
It's strange to return a table with just the ID, length and offset. The old API took the x, y, and timestamps and sorted them by the ID and timestamp.  I don't think that the gather required for this re-sorting should be moved outside of the library.

In fact, if this function only returns the *unique* IDs, then how can you use it to reorder the x/y/timestamp data?

I think it's probably best to produce the sorted data in this function. You could return a pair of tables: one with the trajectory ID/length/offset, and one with the sorted dataset. Also, ID is not used in any of the other APIs, once the unique IDs are computed. So is it necessary to return it? 

Length is not strictly needed either; if you make the ID/offset table one larger, then you can compute the lengths on the fly. If length is not needed outside of cuSpatial that would save some precious memory and bandwidth.

So it could be possible to just return a cudf::column of offsets and the sorted data, and that would be the input to both of the other trajectory functions.

Reply:
Upon closer inspection, I think we might be able to eliminate the `derive_trajectories` groupby altogether, and instead implement `compute_velocities` and `compute_bounding_boxes` via `thrust::reduce_by_key` on the unsorted data.

Reply:
Yeah it seems those extra columns were redundant, which eliminates the need to use groupby. For that reason, and the fact we can't get the groupby offsets as a device buffer, I've changed it back to a thrust reduce_by_key over the sorted input ids.